### PR TITLE
Fix #5054: Remove the dead code generated by the Labeled-based pattern matcher

### LINKED ISSUE #5054: Unnecessary tree generated by PatternMatcher
While compiling something like:
```scala
class Test {
  def test(s: String) = s match {
    case "Hello" => 1
    case _       => 2
  }
}
```
after `PatternMatcher`, we get a tree of the form:
```scala
class Test {
  def test(s: String): Int =
    matchResult1[Int]:
      {
        case val x1: String = s
        if "Hello".==(x1) then return[matchResult1] 1 else ()
        return[matchResult1] 2
        return[matchResult1] throw new MatchError(x1)
      }
}
```
It seems unnecessary to generate the tree `return[matchResult1] throw new MatchError(x1)`.

Before the labelled block PR:
```scala
class Test {
  def test(s: String): Int = {
    case val x1: String(s) = s
    if "Hello".==(x1) then 1 else 2
  }
}
```
Also unnecessary bytecode seems to be generated:
```
Compiled from "Test.scala"
public class Test {
  public int test(java.lang.String);
    Code:
       0: aload_1
       1: astore_2
       2: ldc           #15                 // String Hello
       4: aload_2
       5: invokevirtual #19                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
       8: ifeq          18
      11: iconst_1
      12: goto          34
      15: nop
      16: nop
      17: athrow
      18: iconst_2
      19: goto          34
      22: nop
      23: nop
      24: nop
      25: nop
      26: nop
      27: nop
      28: nop
      29: nop
      30: nop
      31: nop
      32: nop
      33: athrow
      34: ireturn

```
Before the labelled block PR:
```
Compiled from "Test.scala"
public class Test {
  public int test(java.lang.String);
    Code:
       0: aload_1
       1: astore_2
       2: ldc           #15                 // String Hello
       4: aload_2
       5: invokevirtual #19                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
       8: ifeq          15
      11: iconst_1
      12: goto          16
      15: iconst_2
      16: ireturn
}
```

### PR DESCRIPTION
**Pull request by sjrd**
After this PR, the tree after patmat is
```scala
    def test(s: String): Int =
      matchResult1[Int]:
        {
          case val x1: String = s
          if "Hello".==(x1) then return[matchResult1] 1 else ()
          return[matchResult1] 2
        }
```
and the bytecode is
```
  public int test(java.lang.String);
    Code:
       0: aload_1
       1: astore_2
       2: ldc           #15                 // String Hello
       4: aload_2
       5: invokevirtual #19                 // Method java/lang/Object.equals:(Ljava/lang/Object;)Z
       8: ifeq          15
      11: iconst_1
      12: goto          19
      15: iconst_2
      16: goto          19
      19: ireturn
```
The only difference wrt. the previous implementation is now the useless `goto 19` at offset 16. But that should be trivial for a peephole optimizer to get rid of.

### CODE REVIEW

**Comment by allanrenucci** [comment]
>After this PR, the tree after patmat is
```scala
    def test(s: String): Int =
      matchResult1[Int]:
        {
          case val x1: String = s
          if "Hello".==(x1) then return[matchResult1] 1 else ()
          return[matchResult1] 2
        }
```

I see that you added a special case in the backend to not generate code if the else part if the `Unit` constant. Can you not generate the else branch of the `if` in the first place?

Also I think it would be good to add a bytecode test, that make sure that no `athrow` instruction is generated:
```scala
  @Test def labeledBlock = {
    val source =
      """class Test {
        |  def test(s: String) = s match {
        |    case "Hello" => 1
        |    case _       => 2
        |  }
        |}
      """.stripMargin

    checkBCode(source) { dir =>
      val clsIn = dir.lookupName("Test.class", directory = false)
      val clsNode = loadClassNode(clsIn.input)
      val method = getMethod(clsNode, "test")
      val throwMatchError = instructionsFromMethod(method).exists {
        case Op(Opcodes.ATHROW) => true
        case _ => false
      }
      assertFalse(throwMatchError)
    }
  }
```

**Comment by sjrd** [comment]
> Can you not generate the else branch of the if in the first place?

I'm afraid not. dotty internals do not like that. In dotty a `tpd.If` must have an `elsep`, and it cannot be `EmptyTree` (I tried, things blow up everywhere).

**Comment by sjrd** [comment]
> Also I think it would be good to add a bytecode test, that make sure that no athrow instruction is generated:

I have added the test, thanks! As well as a few other tests for switches.

File: compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala [hunk 1]
```diff
@@ -597,6 +603,16 @@ object PatternMatcher {
             plan
           }
         }
+        override def apply(plan: SeqPlan): Plan = {
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala]: You should update the doc comment of `def inlineVars` saying that it drops unreachable branch as well

**Comment by sjrd** [comment]
Updated. Rebased on master and addressed @allanrenucci's comments.

**Pull request merged**